python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -181,6 +222,14 @@ def end_training():
 
             return output
 
+        futures = self.worker_group.execute_async(pause_reporting)
+        self.get_with_failure_handling(futures)
+
+        # Finish up processing results.
+        results = self.fetch_next_result()
+        while results is not None:
+            results = self.fetch_next_result()
```

Comment:
Actually maybe instead of calling `fetch_next_result` in here it might make sense to separate out the checkpoint processing into its own thing which can be called both in `fetch_next_result` and `finish_training`.

Because here, if we disable reporting, then `fetch_next_result` is not actually returning any results. It's just being used to process checkpoints. Also that way, we should be able to avoid the recursion in `fetch_next_result`.

Reply:
Good call. I extracted out the common logic into a `_get_next_results` method and updated both `fetch_next_result` and `finish_training` with their own iterative processing logic. Please take a look!